chia/full_node/weight_proof_v2.py

```diff
@@ -0,0 +1,1837 @@
+    prev_challenge_ip_iters = uint64(0)
+    slot_after_challenge = False
+    for idx, segment in enumerate(segments):
+        log.debug(f"validate segment {idx} sampled:{sampled_seg_index == idx}")
+        res = _validate_segment(
+            constants,
+            segment,
+            curr_ssi,
+            curr_difficulty,
+            prev_ses if idx == 0 else None,
+            sampled_seg_index == idx,
+            cc_sub_slot_hash,
+            icc_sub_slot_hash,
+            uint64(0) if slot_after_challenge else prev_challenge_ip_iters,
+            start_idx,
+        )
+        prev_challenge_ip_iters, slot_iters, slots, cc_sub_slot_hash, icc_sub_slot_hash, slot_after_challenge = res
+        log.debug(f"cc sub slot hash {cc_sub_slot_hash}")
+        if prev_ses is not None and prev_ses.new_sub_slot_iters is not None:
+            curr_ssi = prev_ses.new_sub_slot_iters
+        if prev_ses is not None and prev_ses.new_difficulty is not None:
+            curr_difficulty = prev_ses.new_difficulty
+
+        total_blocks += 1
+        total_slot_iters += slot_iters * slots
+        total_slots += slots
+        total_ip_iters += prev_challenge_ip_iters
+    avg_ip_iters = total_ip_iters / total_blocks
+    avg_slot_iters = total_slot_iters / total_slots
+    if avg_slot_iters / avg_ip_iters < float(constants.WEIGHT_PROOF_THRESHOLD):
+        raise Exception(
+            f"bad avg challenge block positioning ratio: {avg_slot_iters / avg_ip_iters} sub_epoch {sub_epoch_n}"
+        )
+    return sub_epoch_n
+
+
+def _validate_segment(
+    constants: ConsensusConstants,
+    segment: SubEpochChallengeSegmentV2,
+    curr_ssi: uint64,
+    curr_difficulty: uint64,
+    ses: Optional[SubEpochSummary],
+    sampled: bool,
+    cc_challenge: bytes32,
+    icc_challenge: Optional[bytes32],
+    prev_challenge_ip_iters: uint64,
+    start_from: int = 0,
+) -> Tuple[uint64, uint64, int, bytes32, bytes32, bool]:
+    """
+    validates all segments
+    if this is the "sampled" challenge segment we validate all the vdfs
+    if not we validate all the hashes for the end of slots
+
+    start_from is used for an edge case where this is the first segment
+    of the sub epoch and there are multiple empty slots
+    in that case we already have have the end of slot challenges so we can skip those first slots
+
+    when we validate vdfs with compressed values we get the uncompressed value as a result
+    output_cache is used to store all these converted B objects to use as inputs for later vdfs
+    """
+
+    slot_iters, slots = uint64(0), 0
+    output_cache: Dict[B, ClassgroupElement] = {}
+    sub_slot_data = segment.sub_slot_data[start_from:]
+    first_block = True
+    prev_cc_challenge = None
+    after_challenge_block = False
+    slot_after_challenge_block = False
+    deficit = 0
+    for idx, ssd in enumerate(sub_slot_data):
+        if ssd.is_challenge():
```

Comment:
What happens if none of them are challenge? Shouldn't we check that there are at least 12 challenge blocks per subepoch?

Reply:
we can add a check counting the number of challenges, anything out of place here would cause the vdf and hash validation to fail since everything depends on the challenges